cmd2/rl_utils.py

```diff
@@ -85,14 +130,73 @@ def rl_force_redisplay() -> None:
     if not sys.stdout.isatty():
         return
 
-    if rl_type == RlType.GNU:  # pragma: no cover
+    if rl_type == RlType.GNU:
         readline_lib.rl_forced_update_display()
 
         # After manually updating the display, readline asks that rl_display_fixed be set to 1 for efficiency
         display_fixed = ctypes.c_int.in_dll(readline_lib, "rl_display_fixed")
         display_fixed.value = 1
 
-    elif rl_type == RlType.PYREADLINE:  # pragma: no cover
+    elif rl_type == RlType.PYREADLINE:
         # Call _print_prompt() first to set the new location of the prompt
         readline.rl.mode._print_prompt()
         readline.rl.mode._update_line()
+
+
+# noinspection PyProtectedMember
+def rl_get_point() -> int:  # pragma: no cover
+    """
+    Returns the offset of the current cursor position in rl_line_buffer
+    """
+    if rl_type == RlType.GNU:
+        return ctypes.c_int.in_dll(readline_lib, "rl_point").value
+
+    elif rl_type == RlType.PYREADLINE:
+        return readline.rl.mode.l_buffer.point
+
+    else:
+        return 0
+
+
+# noinspection PyProtectedMember
+def rl_set_prompt(prompt: str) -> None:  # pragma: no cover
+    """
+    Sets readline's prompt
+    :param prompt: the new prompt value
+    """
+    safe_prompt = rl_make_safe_prompt(prompt)
+
+    if rl_type == RlType.GNU:
+        encoded_prompt = bytes(safe_prompt, encoding='utf-8')
+        readline_lib.rl_set_prompt(encoded_prompt)
+
+    elif rl_type == RlType.PYREADLINE:
+        readline.rl._set_prompt(safe_prompt)
+
+
+def rl_make_safe_prompt(prompt: str, start: str = "\x01", end: str = "\x02") -> str:  # pragma: no cover
+    """Overcome bug in GNU Readline in relation to calculation of prompt length in presence of ANSI escape codes.
+
+    :param prompt: original prompt
+    :param start: start code to tell GNU Readline about beginning of invisible characters
+    :param end: end code to tell GNU Readline about end of invisible characters
+    :return: prompt safe to pass to GNU Readline
+    """
+    if rl_type == RlType.GNU:
+        escaped = False
+        result = ""
+
+        for c in prompt:
+            if c == "\x1b" and not escaped:
+                result += start + c
+                escaped = True
+            elif c.isalpha() and escaped:
```

Comment:
It would be nice if I knew a little more about the ANSI escape codes in order to understand why encountering a letter means the escape code is over

Reply:
It would be nice, for all of us, if you knew that Lance :)

Reply:
Yes, this is one of the great mysteries of computing that has been lost to the mists of time.

Reply:
Okay, I learned a little more. It looks like all escape codes end with a character in the range 0x40 to 0x7e. So while this check may be insufficient to know that any escape code has ended, perhaps it's good enough for cmd2 code because we usually only have color codes in our prompts, all of which end with 'm' (0x6d).